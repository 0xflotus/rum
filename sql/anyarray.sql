set enable_seqscan=off;
SET enable_bitmapscan=OFF;

CREATE TABLE test_anyarray(
	i int[]
);
INSERT INTO test_anyarray
SELECT ARRAY[1, x, x + 1] FROM generate_series(1, 10000) x;

CREATE INDEX test_anyarray_idx ON test_anyarray USING rum (i);

EXPLAIN (costs off)
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i % '{1}' ORDER BY i <=> '{1}' LIMIT 5;
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i % '{1}' ORDER BY i <=> '{1}' LIMIT 5;

EXPLAIN (costs off)
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i @> '{1}' ORDER BY i <=> '{1}' LIMIT 5;
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i @> '{1}' ORDER BY i <=> '{1}' LIMIT 5;

EXPLAIN (costs off)
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i % '{1}' ORDER BY i <=> '{1}' LIMIT 5000;

CREATE TABLE test_anyarray_ts(
	i int[],
	d timestamp
);
INSERT INTO test_anyarray_ts VALUES
	( '{1,2,3}', '2004-10-26 03:55:08' ),
	( '{2,3,4}', '2004-10-26 04:55:08' ),
	( '{3,4,5}', '2004-10-26 05:55:08' ),
	( '{4,5,6}', '2004-10-26 08:55:08' ),
	( '{5,6,7}', '2004-10-26 09:55:08' ),
	( '{6,7,8}', '2004-10-26 10:55:08' )
;

CREATE INDEX test_anyarray_ts_idx ON test_anyarray_ts USING rum
	(i rum_anyarray_addon_ops, d)
	WITH (attach = 'd', to = 'i');

EXPLAIN (costs off)
SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i % '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i % '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;

EXPLAIN (costs off)
SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i @> '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i @> '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
