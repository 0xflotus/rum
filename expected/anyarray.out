set enable_seqscan=off;
SET enable_bitmapscan=OFF;
CREATE TABLE test_anyarray(
	i int[]
);
INSERT INTO test_anyarray
SELECT ARRAY[1, x, x + 1] FROM generate_series(1, 10000) x;
CREATE INDEX test_anyarray_idx ON test_anyarray USING rum (i);
EXPLAIN (costs off)
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i % '{1}' ORDER BY i <=> '{1}' LIMIT 5;
                        QUERY PLAN                         
-----------------------------------------------------------
 Limit
   ->  Index Scan using test_anyarray_idx on test_anyarray
         Index Cond: (i % '{1}'::integer[])
         Order By: (i <=> '{1}'::integer[])
(4 rows)

SELECT *, i <=> '{1}' FROM test_anyarray WHERE i % '{1}' ORDER BY i <=> '{1}' LIMIT 5;
    i    |     ?column?     
---------+------------------
 {1,1,2} |  1.4142135623731
 {1,2,3} | 1.73205080756888
 {1,3,4} | 1.73205080756888
 {1,4,5} | 1.73205080756888
 {1,5,6} | 1.73205080756888
(5 rows)

EXPLAIN (costs off)
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i @> '{1}' ORDER BY i <=> '{1}' LIMIT 5;
                        QUERY PLAN                         
-----------------------------------------------------------
 Limit
   ->  Index Scan using test_anyarray_idx on test_anyarray
         Index Cond: (i @> '{1}'::integer[])
         Order By: (i <=> '{1}'::integer[])
(4 rows)

SELECT *, i <=> '{1}' FROM test_anyarray WHERE i @> '{1}' ORDER BY i <=> '{1}' LIMIT 5;
    i    |     ?column?     
---------+------------------
 {1,1,2} |  1.4142135623731
 {1,2,3} | 1.73205080756888
 {1,3,4} | 1.73205080756888
 {1,4,5} | 1.73205080756888
 {1,5,6} | 1.73205080756888
(5 rows)

EXPLAIN (costs off)
SELECT *, i <=> '{1}' FROM test_anyarray WHERE i % '{1}' ORDER BY i <=> '{1}' LIMIT 5000;
                        QUERY PLAN                         
-----------------------------------------------------------
 Limit
   ->  Index Scan using test_anyarray_idx on test_anyarray
         Index Cond: (i % '{1}'::integer[])
         Order By: (i <=> '{1}'::integer[])
(4 rows)

CREATE TABLE test_anyarray_ts(
	i int[],
	d timestamp
);
INSERT INTO test_anyarray_ts VALUES
	( '{1,2,3}', '2004-10-26 03:55:08' ),
	( '{2,3,4}', '2004-10-26 04:55:08' ),
	( '{3,4,5}', '2004-10-26 05:55:08' ),
	( '{4,5,6}', '2004-10-26 08:55:08' ),
	( '{5,6,7}', '2004-10-26 09:55:08' ),
	( '{6,7,8}', '2004-10-26 10:55:08' )
;
CREATE INDEX test_anyarray_ts_idx ON test_anyarray_ts USING rum
	(i rum_anyarray_addon_ops, d)
	WITH (attach = 'd', to = 'i');
EXPLAIN (costs off)
SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i % '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
                                    QUERY PLAN                                     
-----------------------------------------------------------------------------------
 Limit
   ->  Index Scan using test_anyarray_ts_idx on test_anyarray_ts
         Order By: (d <=> 'Tue Oct 26 08:00:00 2004'::timestamp without time zone)
         Filter: (i % '{3}'::integer[])
(4 rows)

SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i % '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
    i    |            d             | ?column? 
---------+--------------------------+----------
 {3,4,5} | Tue Oct 26 05:55:08 2004 |     7492
 {2,3,4} | Tue Oct 26 04:55:08 2004 |    11092
 {1,2,3} | Tue Oct 26 03:55:08 2004 |    14692
(3 rows)

EXPLAIN (costs off)
SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i @> '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
                                    QUERY PLAN                                     
-----------------------------------------------------------------------------------
 Limit
   ->  Index Scan using test_anyarray_ts_idx on test_anyarray_ts
         Index Cond: (i @> '{3}'::integer[])
         Order By: (d <=> 'Tue Oct 26 08:00:00 2004'::timestamp without time zone)
(4 rows)

SELECT *, d <=> '2004-10-26 08:00:00' FROM test_anyarray_ts WHERE i @> '{3}' ORDER BY d <=> '2004-10-26 08:00:00' LIMIT 5;
    i    |            d             | ?column? 
---------+--------------------------+----------
 {3,4,5} | Tue Oct 26 05:55:08 2004 |     7492
 {2,3,4} | Tue Oct 26 04:55:08 2004 |    11092
 {1,2,3} | Tue Oct 26 03:55:08 2004 |    14692
(3 rows)

